### PR TITLE
[Rundeck] Support for shouldIncludeRundeckLogs

### PR DESCRIPTION
**Pull request by szpak**
As shouldIncludeRundeckLogs requires shouldFailTheBuild to be set I also added a shortcut `shouldWaitForRundeckJobAndIncludeRundeckLogs`. It is convenient, but I don't know how it fits into job-dsl conventions.

### CODE REVIEW

**Comment by daspilker** [comment]
How can I test this without a running instance of Rundeck?

**Comment by szpak** [comment]
In case it would be enough for you to verify that the same parameters are passed to a Rundeck instance you could compare XML created with manual marking that checkbox in Jenkins job configuration with the XML generated from Jenkins Job DSL with my change.
I have to admit I haven't test the Job DSL plugin version with my change and I end automatic testing on PublisherContextSpec which suggests that proper nodes are created in generated XML. How do you usually test the new features like that?

Regarding manual functional tests with Rundeck in my current build I have the following code and it works :).

```
configure { projectNode ->
    projectNode / 'publishers' / 'org.jenkinsci.plugins.rundeck.RundeckNotifier' << {
        'includeRundeckLogs'('true')
    }
}
```

**Comment by daspilker** [comment]
I always do some manual testing in Jenkins to ensure that there are no type in element names. I could not save the job configuration page because saving already seems to require a running instance of Rundeck.

**Comment by szpak** [comment]
Rundeck is a jar which can be run without installation. It is only required to create project and job and generate access token, but If you prefer I can send you a fragment of XML created with manual clicking in Jenkins.

File: docs/Job-reference.md [hunk 1]
```diff
@@ -5154,13 +5154,15 @@ job('example-2') {
...
-            nodeFilters(Map<String, String> filters)      // defaults to empty map if omitted
+            tag(String tag)                                 // defaults to empty string if omitted
+            nodeFilters(Map<String, String> filters)        // defaults to empty map if omitted
             nodeFilter(String key, String value)
-            shouldWaitForRundeckJob(boolean value = true) // defaults to false if omitted
-            shouldFailTheBuild(boolean value = true)      // defaults to false if omitted
+            shouldWaitForRundeckJob(boolean value = true)   // defaults to false if omitted
+            shouldFailTheBuild(boolean value = true)        // defaults to false if omitted
+            shouldIncludeRundeckLogs(boolean value = true)  // defaults to false if omitted - since 1.35
+            shouldWaitForRundeckJobAndIncludeRundeckLogs(boolean value = true)  // sets both options as shouldIncludeRundeckLogs requires shouldWaitForRundeckJob - since 1.35
```
Comment (daspilker) [inline docs/Job-reference.md]: do not include shortcut methods, just set shouldWaitForRundeckJob when setting shouldIncludeRundeckLogs

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/RundeckContext.groovy [hunk 2]
```diff
@@ -36,4 +37,21 @@ class RundeckContext implements Context {
     void shouldFailTheBuild(boolean shouldFailTheBuild = true) {
         this.shouldFailTheBuild = shouldFailTheBuild
     }
+
+    //Requires Rundeck plugin 3.4+
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/RundeckContext.groovy]: use the @RequiresPlugin annotation

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext.groovy [hunk 3]
```diff
@@ -1036,6 +1036,7 @@ class PublisherContext extends AbstractExtensibleContext {
             tag rundeckContext.tag
             shouldWaitForRundeckJob rundeckContext.shouldWaitForRundeckJob
             shouldFailTheBuild rundeckContext.shouldFailTheBuild
+            shouldIncludeRundeckLogs rundeckContext.shouldIncludeRundeckLogs
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext.groovy]: This element should only be included when version 3.4 or later of the rundeck plugin is installed

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/RundeckContext.groovy (hunk 2 above)
Comment (szpak) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/RundeckContext.groovy]: First I tried (before changing it into a comment), but I've got:
```
startup failed:
       @RequiresPlugin(id = 'rundeck', minimumVersion = '3.4')
       ^
/.../job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/RundeckContext.groovy: 42: no jobManagement field in javaposse.jobdsl.dsl.helpers.publisher.RundeckContext
```
Can that annotation be used with a method like that?

Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/RundeckContext.groovy]: As the messages says, you need a jobManagement field in your class because the annotation needs a JobManagement instance to perform the check. You get that when the context inherits from AbstractContext.

Comment (szpak) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/RundeckContext.groovy]: Ok, I can add it, but the way making my modifications.

**Comment by szpak** [comment]
I reworked the code after your comments.

**Comment by szpak** [comment]
By the way I also simplified a few tests in the neighborhood.

**Comment by burdandrei** [comment]
This one is good! :+1:

File: job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy [hunk 4]
```diff
@@ -2950,7 +2991,7 @@ class PublisherContextSpec extends Specification {
         'test' | 'test' | null
     }
 
-    def 'call s3 with invalid storage class'(String storageClass) {
+    def 'call s3 with invalid storage class'() {
```
Comment (daspilker) [inline job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy]: do not fix unrelated tests, open a separate pull request for that

**Comment by daspilker** [comment]
This does not work in Jenkins, the checkbox is not set. Test this in Jenkins and fix the PR.

**Comment by szpak** [comment]
In the end there was no technical error in code, but Rundeck has inconsistently named variables and I had `shouldIncludeRundeckLogs` instead of `includeRundeckLogs`. I tested it manually with Jenkins 1.609.2.

Is there something else I should update in regards to recent documentation changes?

**Comment by daspilker** [comment]
Add GroovyDoc for the new DSL method. And so not refactor unrelated tests, keep the diff at a minimum.

**Comment by szpak** [comment]
Documentation moved from markdown file to GroovyDoc.

I reverted my changes to unrelated test in the previous commit... except one which was next to rundeck tests (which I missed). Currently only Rundeck related tests are modified.

File: job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy [hunk 5]
```diff
@@ -3042,21 +3043,27 @@ class PublisherContextSpec extends Specification {
             tag 'tag'
             shouldWaitForRundeckJob()
             shouldFailTheBuild false
+            shouldIncludeRundeckLogs()
         }
 
         then:
         Node rundeckNode = context.publisherNodes[0]
         rundeckNode.name() == 'org.jenkinsci.plugins.rundeck.RundeckNotifier'
-        rundeckNode.jobId[0].value() == 'jobId'
```
Comment (daspilker) [inline job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy]: revert these changes, no need to use the safe navigation operator here

File: job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy [hunk 6]
```diff
@@ -3042,21 +3043,27 @@ class PublisherContextSpec extends Specification {
...
+        rundeckNode.shouldWaitForRundeckJob[0]?.value() == true
+        rundeckNode.shouldFailTheBuild[0]?.value() == false
+        rundeckNode.includeRundeckLogs[0]?.value() == true
+
+        and:
         1 * jobManagement.requirePlugin('rundeck')
+        1 * jobManagement.requireMinimumPluginVersion('rundeck', '3.4')
     }
 
-    def 'call rundeck with invalid jobId should fail'() {
```
Comment (daspilker) [inline job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy]: revert these changes

File: job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy [hunk 7]
```diff
@@ -3065,25 +3072,57 @@ class PublisherContextSpec extends Specification {
         exception.message =~ /\(.+, line \d+\) jobIdentifier cannot be null or empty/
 
         where:
-        id   | _
```
Comment (daspilker) [inline job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy]: and revert these changes

File: job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy [hunk 8]
```diff
@@ -3065,25 +3072,57 @@ class PublisherContextSpec extends Specification {
...
     def 'call rundeck with default values'() {
+        given:
+        jobManagement.getPluginVersion('rundeck') >> new VersionNumber('3.4')
+
         when:
         context.rundeck('jobId')
 
         then:
         Node rundeckNode = context.publisherNodes[0]
-        rundeckNode.options[0].value().isEmpty()
```
Comment (daspilker) [inline job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy]: revert these changes, no need to use the safe navigation operator here

File: job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy (hunk 5 above)
Comment (szpak) [inline job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy]: During the new feature development I came across a situation (after I renamed a node name) when assertion like that failed. The error message was:

```
java.lang.NullPointerException: Cannot invoke method value() on null object
```

From the error message I know that there wasn't node with given name, but I cannot say why - developer has to debug it or add some additional logging statement before. 

After my change the error message became much more meaningful:
```
rundeckNode.includeRundeckLogsWrongName[0]?.value() == true
|           |                  |    |       |
|           []                 null null    false
org.jenkinsci.plugins.rundeck.RundeckNotifier[attributes={}; value=[jobId[attributes={}; value=jobId],
options[attributes={}; value=], nodeFilters[attributes={}; value=], tag[attributes={}; value=], 
shouldWaitForRundeckJob[attributes={}; value=true], shouldFailTheBuild[attributes={}; value=false],
includeRundeckLogs[attributes={}; value=true]]]
```

It is in some sense a boiler plate code, but unfortunately currently it [doesn't support](https://github.com/spockframework/spock/pull/49) it. I personally prefer a need to add additional `?` to get verbose error message in case of regression.

File: job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy (hunk 6 above)
Comment (szpak) [inline job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy]: I reverted changes in all not Rundeck related tests, but as I made modifications in Rundeck code I would like to keep Rundeck related tests in a good fashion in case of regression in the future. Without `@Unroll` It is harder to see for which input data the condition failed.

File: job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy (hunk 7 above)
Comment (szpak) [inline job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy]: Data tables are a nice syntactic sugar for data pipes, but it doesn't look very readable for just a single variables.

File: job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy (hunk 8 above)
Comment (szpak) [inline job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy]: The same explanation as before.

**Comment by kamilszymanski** [comment]
@daspilker any plans for merging this PR?
@szpak you need to rebase this PR - there are conflicts to be resolved

**Comment by szpak** [comment]
I rebased my PR - there are still conflicts in HOME.md.

I already removed all non rundeck related tests as requested by @daspilker.
